stellargraph/core/convert.py

```diff
@@ -150,3 +154,144 @@ def convert_edges(
 
     return EdgeData(edges, node_data)
+
+
+NodeInfo = namedtuple("NodeInfo", ["ids", "features"])
+
+
+def _empty_node_info() -> NodeInfo:
+    return NodeInfo([], [])
+
+
+def _features_from_attributes(node_type, num_nodes, values, dtype):
+    # the size is the first element that has a length, or None if there's only None elements.
+    size = next((len(x) for x in values if x is not None), None)
+
+    if size is None:
+        # no features = zero-dimensional features, and skip the loop below
+        return np.empty((num_nodes, 0), dtype)
+
+    default_value = np.zeros(size, dtype)
+
+    def compute_value(x):
+        if x is None:
+            return default_value
+        elif len(x) != size:
+            raise ValueError(
+                f"inferred all nodes of type {node_type!r} to have feature dimension {size}, found dimension {len(x)}"
+            )
+
+        return x
+
+    matrix = np.array([compute_value(x) for x in values], dtype)
+    assert matrix.shape == (num_nodes, size)
+
+    return matrix
+
+
+def _features_from_node_data(nodes, data, dtype):
+    if isinstance(data, dict):
+
+        def single(node_type):
+            node_info = nodes[node_type]
+            this_data = data[node_type]
+
+            if isinstance(this_data, pd.DataFrame):
+                df = this_data.astype(dtype, copy=False)
+            elif isinstance(this_data, (Iterable, list)):
+                ids, values = zip(*this_data)
+                df = pd.DataFrame(values, index=ids, dtype=dtype)
+
+            graph_ids = set(node_info.ids)
+            data_ids = set(df.index)
+            if graph_ids != data_ids:
+                parts = []
+                missing = graph_ids - data_ids
+                if missing:
+                    parts.append(f"missing from data ({comma_sep(missing)})")
+                extra = data_ids - graph_ids
+                if extra:
+                    parts.append(f"extra in data ({comma_sep(extra)})")
+                message = " and ".join(parts)
+                raise ValueError(
+                    f"expected feature node IDs to exactly match nodes in graph; found: {message}"
+                )
+
+            return df
+
+        return {node_type: single(node_type) for node_type in nodes.keys()}
+    elif isinstance(data, pd.DataFrame):
+        if len(nodes) > 1:
+            raise TypeError(
+                "When there is more than one node type, pass node features as a dictionary."
+            )
+
+        node_type = next(iter(nodes))
+        return _features_from_node_data(nodes, {node_type: data}, dtype)
+    elif isinstance(data, (Iterable, list)):
+        id_to_data = dict(data)
+        return {
+            node_type: pd.DataFrame(
+                (id_to_data[x] for x in node_info.ids), index=node_info.ids, dtype=dtype
+            )
+            for node_type, node_info in nodes.items()
+        }
+
+
+def _fill_or_assign(df, column, default):
```

Comment:
Only generally trying to avoid mutations, however this is a mutation of a value constructed by this code and not otherwise exposed, so mutation is fine. Good catch 👍

Reply:
one minor optional thing - `_fill_or_assign` no longer uses `assign`. It might be more appropriate to rename it `_fill_or_append`. but then again this a hidden function called by a backend function so up to you :)

Reply:
Ah, yeah, I didn't adjust the name, but in this case I think the assignment of `=` (i.e. assign a new column) is a more natural way to phrase this than "appending to the list of columns". Typically I'd think of append as "add new row(s)" for a dataframe.